src/pages/resultsView/download/DownloadTab.tsx

```diff
@@ -1328,29 +1249,188 @@ export default class DownloadTab extends React.Component<
 
     @autobind
     private handleOtherMolecularProfileDownload(profileName: string) {
-        onMobxPromise(
-            this.allOtherMolecularProfileDownloadDataGroupByProfileName,
-            downloadDataGroupByProfileName => {
-                const textMap = this.downloadDataTextGroupByKey(
-                    downloadDataGroupByProfileName
+        onMobxPromise<any>(
+            [
+                this.props.store.nonSelectedDownloadableMolecularProfiles,
+                this.props.store.samples,
+                this.props.store.genes,
+            ],
+            async (
+                nonSelectedDownloadableMolecularProfiles,
+                samples,
+                genes
+            ) => {
+                // STEP 1: fetch data
+                const selectedProfiles: MolecularProfile[] = nonSelectedDownloadableMolecularProfiles.filter(
+                    (profile: MolecularProfile) => profile.name === profileName
+                );
+
+                let molecularData: any[] = [];
+                if (
+                    selectedProfiles.length &&
+                    genes != undefined &&
+                    genes.length
+                ) {
+                    const selectedProfilesGroupByStudyId = _.groupBy(
+                        selectedProfiles,
+                        profile => profile.studyId
+                    );
+                    // find samples which share studyId with profile and add identifier
+                    const sampleIdentifiers: SampleMolecularIdentifier[] = (samples as Sample[])
+                        .filter(
```

Comment:
there are a couple easy optimizations here.  1. you don't need to filter first.  you can just put a condition in the reducer that just returns acc if studyId doesn't appear in selected profile collection.   
2. you should use acc.push instead of acc.concat.    concat produces a new array each time. so, for every sample, you're producing a new array.  push changes array in place, which is find here.

Reply:
Good to know the second one, thanks! I will change that.